File: Week_06-动态规划/121.买卖股票的最佳时机.py
#
# @lc app=leetcode.cn id=121 lang=python
#
# [121] 买卖股票的最佳时机
#
'''
定义状态dp[i][j]是第i天手里的现金数目
dp[i][0]：第i天，不持有股票时，手里的现金
dp[i][1]：第i天，持有股票时，手里的现金

#注意和122的差别在于，此题目中不能多次买卖股票，因此当前持有股票的话，一种是原来持有股票当前继续持有
另外一种是，原来没买过，今天买下股票:0-prices[i]
不可能的情况是多次买卖：dp[i][0]-prices[i]


'''


# @lc code=start
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        dp = [[0]*2]*(n)
        dp[0][0] = 0
        dp[0][1] = -prices[0]
        for i in range(1, n):
            dp[i][0] = max(dp[i-1][0], dp[i-1][1]+prices[i])
            dp[i][1] = max(dp[i-1][1], -prices[i])  
        return dp[n-1][0]



# @lc code=end

File: Week_06-动态规划/121.买卖股票的最佳时机.cpp
/*
 * @lc app=leetcode.cn id=121 lang=cpp
 *
 * [121] 买卖股票的最佳时机
 */
/*
买卖股票问题，定义状态：
dp[i][k][b]第i天，买入k次，手上有股票或无股票时的利润
如果不考虑买入次数，则不需要定义k维度。
注意只能买入一次股票，就是k<=1
*/

// @lc code=start
class Solution {
public:
    int maxProfit(vector<int>& prices) {
#if 0
        //定义状态数组的写法
        int n =prices.size();
        if (n==0) return 0;
        vector<vector<int>> dp(n, vector<int>(2,0));
        for (int i=0; i<n; i++) {
            if(i==0) {
                dp[i][0] = 0;
                dp[i][1] = -prices[i];
                continue;
            }
            dp[i][0] = max(dp[i-1][0], dp[i-1][1]+prices[i]);//今天手上不持有股票：之前一天也不持有，之前一天持有但是今天卖掉了
            
            dp[i][1] = max(dp[i-1][1], -prices[i]);  //今天手上持有股票：之前买入过，或者是之前没买过（利润为0）今天买入。
        }
        return dp[n-1][0];
#endif
#if 1
//降低空间复杂度的写法
//注意状态转移方程，新状态只和相邻状态相关，因此可以把数组改为变量，降低空间复杂度。
        int n = prices.size();
        if (n==0) return 0;
        int dp_i_0 = 0;
        int dp_i_1 = -prices[0];
        for (auto price : prices) {
            dp_i_0 = max(dp_i_1+price, dp_i_0);
            dp_i_1 = max(dp_i_1, -price);
        }
        return dp_i_0;
#endif
    }
};
// @lc code=end

